### ISSUE TITLE
Use private docker registries for toolkit and runtime docker images

### ISSUE THREAD

**Issue by pubudu538** [wso2/k8s-api-operator#338]
**Description:**
There are two use cases here.

1. Private docker registry that is not secured with Basic Auth
This is what Renuka has tested.

2. Private docker registry that is secured with Basic Auth
For this, we need to update the Kaniko job with a docker pull secret.

**Suggested Labels:**

**Suggested Assignees:**

**Affected Product Version:**

**OS, DB, other environment details and versions:**    

**Steps to reproduce:**

**Related Issues:**

**Comment by renuka-fernando** [comment]
**Working fine for HTTP registry secured with Basic Auth (use case 02)**

## Use Case 2: Secured with Basic Auth

### 1.  Create an HTTP registry.
```sh
$ docker run \
  --entrypoint htpasswd \
  registry:2 -Bbn renuka renuka1234 > auth/htpasswd


$ docker run -d \
  -p 5000:5000 \
  --restart=always \
  --name registry-auth \
  -v "$(pwd)"/auth:/auth \
  -e "REGISTRY_AUTH=htpasswd" \
  -e "REGISTRY_AUTH_HTPASSWD_REALM=Registry Realm" \
  -e REGISTRY_AUTH_HTPASSWD_PATH=/auth/htpasswd \
  registry:2
```

### 2. Pushe MGW toolkit and runtime images
> 192.168.8.132 is my ip
```sh
docker login 192.168.8.132:5000

docker tag wso2am/wso2micro-gw-toolkit:3.2.0-alpha 192.168.8.132:5000/wso2am/wso2micro-gw-toolkit:3.2.0-alpha
docker tag wso2/wso2micro-gw:3.2.0-alpha 192.168.8.132:5000/wso2/wso2/wso2micro-gw:3.2.0-alpha

docker push 192.168.8.132:5000/wso2am/wso2micro-gw-toolkit:3.2.0-alpha
docker push 192.168.8.132:5000/wso2/wso2/wso2micro-gw:3.2.0-alpha
```

### 3. Change `controller-config`
```yaml
apiVersion: v1
kind: ConfigMap
metadata:
  name: controller-config
  namespace: wso2-system
data:
  #mgw toolkit image to initialize/setup the micro gw project
  mgwToolkitImg: 192.168.8.132:5000/wso2am/wso2micro-gw-toolkit:3.2.0-alpha
  #mgw runtime image to use in the mgw executable
  mgwRuntimeImg: 192.168.8.132:5000/wso2/wso2/wso2micro-gw:3.2.0-alpha
```

### 4. Install API operator
```sh
$ apictl install api-operator --registry-type=HTTP --repository=192.168.8.132:5000/operator-test --username=renuka --password=renuka1234
```

### 5. Create an API
Scenario 2
```sh
$ apictl add api -n petstore-api --from-file=scenarios/scenario-2/swagger.yaml --override
```

### 6. Teste
```sh
$ kubectl get po
NAME                        READY   STATUS    RESTARTS   AGE
petstore-api-kaniko-9wg4d   1/1     Running   0          9s
```

```sh
$ curl -X GET "https://<external IP of LB service>:9095/petstore/v1/pet/55" -H "accept: application/xml" -H "Authorization:Bearer $TOKEN" -k

<?xml version="1.0" encoding="UTF-8" standalone="yes"?><apiResponse><message>Pet not found</message><type>error</type></apiResponse>
```

**Comment by renuka-fernando** [comment]
Relevant Kaniko Job logs for above use case 02
```sh
INFO[0000] Resolved base name 192.168.8.132:5000/wso2am/wso2micro-gw-toolkit:3.2.0-alpha to 192.168.8.132:5000/wso2am/wso2micro-gw-toolkit:3.2.0-alpha
INFO[0000] Resolved base name 192.168.8.132:5000/wso2/wso2/wso2micro-gw:3.2.0-alpha to 192.168.8.132:5000/wso2/wso2/wso2micro-gw:3.2.0-alpha
INFO[0000] Resolved base name 192.168.8.132:5000/wso2am/wso2micro-gw-toolkit:3.2.0-alpha to 192.168.8.132:5000/wso2am/wso2micro-gw-toolkit:3.2.0-alpha
INFO[0000] Resolved base name 192.168.8.132:5000/wso2/wso2/wso2micro-gw:3.2.0-alpha to 192.168.8.132:5000/wso2/wso2/wso2micro-gw:3.2.0-alpha
INFO[0000] Retrieving image manifest 192.168.8.132:5000/wso2am/wso2micro-gw-toolkit:3.2.0-alpha
INFO[0000] Retrieving image manifest 192.168.8.132:5000/wso2am/wso2micro-gw-toolkit:3.2.0-alpha
INFO[0000] Retrieving image manifest 192.168.8.132:5000/wso2/wso2/wso2micro-gw:3.2.0-alpha
INFO[0000] Retrieving image manifest 192.168.8.132:5000/wso2/wso2/wso2micro-gw:3.2.0-alpha
INFO[0000] Built cross stage deps: map[0:[/home/exec]]
INFO[0000] Retrieving image manifest 192.168.8.132:5000/wso2am/wso2micro-gw-toolkit:3.2.0-alpha
INFO[0000] Retrieving image manifest 192.168.8.132:5000/wso2am/wso2micro-gw-toolkit:3.2.0-alpha
INFO[0000] Unpacking rootfs as cmd RUN micro-gw init project requires it.
INFO[0012] Taking snapshot of full filesystem...
INFO[0013] Resolving paths
INFO[0013] RUN micro-gw init project
INFO[0013] cmd: /bin/sh
INFO[0013] args: [-c micro-gw init project]
JAVA_HOME: /opt/java/openjdk
Initializing Toolkit...
Project 'project' is initialized successfully.

(Use "micro-gw build project" after copying the api definitions)
(Use "micro-gw import project [-l]|[-a -v]" to import APIs from WSO2 API Manager)
INFO[0016] Taking snapshot of full filesystem...
INFO[0016] Resolving paths
INFO[0020] RUN cp /usr/wso2/swagger/project-*/* ./project/api_definitions/
INFO[0020] cmd: /bin/sh
INFO[0020] args: [-c cp /usr/wso2/swagger/project-*/* ./project/api_definitions/]
INFO[0020] Taking snapshot of full filesystem...
INFO[0020] Resolving paths
INFO[0021] RUN cp /usr/wso2/policy/* ./project/
INFO[0021] cmd: /bin/sh
INFO[0021] args: [-c cp /usr/wso2/policy/* ./project/]
INFO[0021] Taking snapshot of full filesystem...
INFO[0021] Resolving paths
INFO[0022] RUN micro-gw build project
INFO[0022] cmd: /bin/sh
INFO[0022] args: [-c micro-gw build project]
JAVA_HOME: /opt/java/openjdk
^C
renuka@renuka scenario-2 % kubectl logs -f petstore-api-kaniko-9wg4d
INFO[0000] Resolved base name 192.168.8.132:5000/wso2am/wso2micro-gw-toolkit:3.2.0-alpha to 192.168.8.132:5000/wso2am/wso2micro-gw-toolkit:3.2.0-alpha
INFO[0000] Resolved base name 192.168.8.132:5000/wso2/wso2/wso2micro-gw:3.2.0-alpha to 192.168.8.132:5000/wso2/wso2/wso2micro-gw:3.2.0-alpha
INFO[0000] Resolved base name 192.168.8.132:5000/wso2am/wso2micro-gw-toolkit:3.2.0-alpha to 192.168.8.132:5000/wso2am/wso2micro-gw-toolkit:3.2.0-alpha
INFO[0000] Resolved base name 192.168.8.132:5000/wso2/wso2/wso2micro-gw:3.2.0-alpha to 192.168.8.132:5000/wso2/wso2/wso2micro-gw:3.2.0-alpha
INFO[0000] Retrieving image manifest 192.168.8.132:5000/wso2am/wso2micro-gw-toolkit:3.2.0-alpha
INFO[0000] Retrieving image manifest 192.168.8.132:5000/wso2am/wso2micro-gw-toolkit:3.2.0-alpha
INFO[0000] Retrieving image manifest 192.168.8.132:5000/wso2/wso2/wso2micro-gw:3.2.0-alpha
INFO[0000] Retrieving image manifest 192.168.8.132:5000/wso2/wso2/wso2micro-gw:3.2.0-alpha
INFO[0000] Built cross stage deps: map[0:[/home/exec]]
INFO[0000] Retrieving image manifest 192.168.8.132:5000/wso2am/wso2micro-gw-toolkit:3.2.0-alpha
INFO[0000] Retrieving image manifest 192.168.8.132:5000/wso2am/wso2micro-gw-toolkit:3.2.0-alpha
INFO[0000] Unpacking rootfs as cmd RUN micro-gw init project requires it.
INFO[0012] Taking snapshot of full filesystem...
INFO[0013] Resolving paths
INFO[0013] RUN micro-gw init project
INFO[0013] cmd: /bin/sh
INFO[0013] args: [-c micro-gw init project]
JAVA_HOME: /opt/java/openjdk
Initializing Toolkit...
Project 'project' is initialized successfully.

(Use "micro-gw build project" after copying the api definitions)
(Use "micro-gw import project [-l]|[-a -v]" to import APIs from WSO2 API Manager)
INFO[0016] Taking snapshot of full filesystem...
INFO[0016] Resolving paths
INFO[0020] RUN cp /usr/wso2/swagger/project-*/* ./project/api_definitions/
INFO[0020] cmd: /bin/sh
INFO[0020] args: [-c cp /usr/wso2/swagger/project-*/* ./project/api_definitions/]
INFO[0020] Taking snapshot of full filesystem...
INFO[0020] Resolving paths
INFO[0021] RUN cp /usr/wso2/policy/* ./project/
INFO[0021] cmd: /bin/sh
INFO[0021] args: [-c cp /usr/wso2/policy/* ./project/]
INFO[0021] Taking snapshot of full filesystem...
INFO[0021] Resolving paths
INFO[0022] RUN micro-gw build project
INFO[0022] cmd: /bin/sh
INFO[0022] args: [-c micro-gw build project]
JAVA_HOME: /opt/java/openjdk
Generating sources...[DONE]

..................................
There are no new updates available
..................................
Compiling source
	wso2/project:3.1.0
ballerina/java.jms:0.8.1 [central.ballerina.io -> home repo]  100% [==========
	ballerina/java.jms:0.8.1 pulled from central successfully

Creating balos
	target/balo/project-2020r1-java8-3.1.0.balo

Running Tests

	wso2/project:3.1.0
	No tests found


Generating executables
	target/bin/project.jar

BUILD SUCCESSFUL
Target: //project/target/project.jar
INFO[0060] Taking snapshot of full filesystem...
INFO[0060] Resolving paths
INFO[0062] RUN mkdir -p /home/exec
INFO[0062] cmd: /bin/sh
INFO[0062] args: [-c mkdir -p /home/exec]
INFO[0062] Taking snapshot of full filesystem...
INFO[0062] Resolving paths
INFO[0063] RUN cp ./project/target/*.jar /home/exec/
INFO[0063] cmd: /bin/sh
INFO[0063] args: [-c cp ./project/target/*.jar /home/exec/]
INFO[0063] Taking snapshot of full filesystem...
INFO[0063] Resolving paths
INFO[0064] Saving file /home/exec for later use
INFO[0064] Deleting filesystem...
INFO[0064] Retrieving image manifest 192.168.8.132:5000/wso2/wso2/wso2micro-gw:3.2.0-alpha
INFO[0064] Retrieving image manifest 192.168.8.132:5000/wso2/wso2/wso2micro-gw:3.2.0-alpha
INFO[0064] Unpacking rootfs as cmd RUN  mkdir -p /home/exec requires it.
INFO[0067] Taking snapshot of full filesystem...
INFO[0067] Resolving paths
INFO[0067] USER root
INFO[0067] cmd: USER
INFO[0067] RUN  mkdir -p /home/exec
INFO[0067] cmd: /bin/sh
INFO[0067] args: [-c mkdir -p /home/exec]
INFO[0067] Taking snapshot of full filesystem...
INFO[0067] Resolving paths
INFO[0068] COPY --from=toolkit /home/exec /home/exec
INFO[0068] Resolving paths
INFO[0068] Taking snapshot of files...
INFO[0068] RUN cp /usr/wso2/mgwconf/* /home/ballerina/conf
INFO[0068] cmd: /bin/sh
INFO[0068] args: [-c cp /usr/wso2/mgwconf/* /home/ballerina/conf]
INFO[0068] Taking snapshot of full filesystem...
INFO[0068] Resolving paths
INFO[0068] RUN echo yes | /home/ballerina/wso2/lib/jdk8*/bin/keytool -import  -storepass ballerina -keystore /home/ballerina/wso2/runtime/bre/security/ballerinaTruststore.p12 -alias "wso2am310-secret-security" -file /usr/wso2/certs/wso2am310-secret/server.pem
INFO[0068] cmd: /bin/sh
INFO[0068] args: [-c echo yes | /home/ballerina/wso2/lib/jdk8*/bin/keytool -import  -storepass ballerina -keystore /home/ballerina/wso2/runtime/bre/security/ballerinaTruststore.p12 -alias "wso2am310-secret-security" -file /usr/wso2/certs/wso2am310-secret/server.pem]
Certificate already exists in keystore under alias <apim300>
Do you still want to add it? [no]:  Certificate was added to keystore
INFO[0069] Taking snapshot of full filesystem...
INFO[0069] Resolving paths
INFO[0069] RUN chmod a+w /home/ballerina
INFO[0069] cmd: /bin/sh
INFO[0069] args: [-c chmod a+w /home/ballerina]
INFO[0069] Taking snapshot of full filesystem...
INFO[0069] Resolving paths
INFO[0070] USER ballerina
INFO[0070] cmd: USER
INFO[0070] ENV project="project"
```

**Comment by renuka-fernando** [comment]
Since in step **4. Install API operator** I selected HTTP registry, the created API image is also pushed to the same HTTP registry that MGW toolkit and runtime has been pushed.

**Comment by renuka-fernando** [comment]
**Working fine for HTTP registry that is not secured with Basic Auth (use case 01)**

## Use Case 1: Not secured with Basic Auth

```sh
$ docker stop registry-auth

$ docker run -d \
  -p 5000:5000 \
  --restart=always \
  --name registry-no-auth \
  registry:2
```

### Follow the same in use case 02
Installation can be done as follows
```sh
$ apictl install api-operator --registry-type=HTTP --repository=192.168.8.132:5000/operator-test
```

### Test

```sh
$ curl -X GET "https://<external IP of LB service>:9095/petstore/v1/pet/55" -H "accept: application/xml" -H "Authorization:Bearer $TOKEN" -k

<?xml version="1.0" encoding="UTF-8" standalone="yes"?><apiResponse><message>Pet not found</message><type>error</type></apiResponse>
```

**Comment by renuka-fernando** [comment]
Relevant Kaniko Job logs for above use case 01
```sh
INFO[0000] Resolved base name 192.168.8.132:5000/wso2am/wso2micro-gw-toolkit:3.2.0-alpha to 192.168.8.132:5000/wso2am/wso2micro-gw-toolkit:3.2.0-alpha
INFO[0000] Resolved base name 192.168.8.132:5000/wso2/wso2/wso2micro-gw:3.2.0-alpha to 192.168.8.132:5000/wso2/wso2/wso2micro-gw:3.2.0-alpha
INFO[0000] Resolved base name 192.168.8.132:5000/wso2am/wso2micro-gw-toolkit:3.2.0-alpha to 192.168.8.132:5000/wso2am/wso2micro-gw-toolkit:3.2.0-alpha
INFO[0000] Resolved base name 192.168.8.132:5000/wso2/wso2/wso2micro-gw:3.2.0-alpha to 192.168.8.132:5000/wso2/wso2/wso2micro-gw:3.2.0-alpha
INFO[0000] Retrieving image manifest 192.168.8.132:5000/wso2am/wso2micro-gw-toolkit:3.2.0-alpha
INFO[0000] Retrieving image manifest 192.168.8.132:5000/wso2am/wso2micro-gw-toolkit:3.2.0-alpha
INFO[0000] Retrieving image manifest 192.168.8.132:5000/wso2/wso2/wso2micro-gw:3.2.0-alpha
INFO[0000] Retrieving image manifest 192.168.8.132:5000/wso2/wso2/wso2micro-gw:3.2.0-alpha
INFO[0000] Built cross stage deps: map[0:[/home/exec]]
INFO[0000] Retrieving image manifest 192.168.8.132:5000/wso2am/wso2micro-gw-toolkit:3.2.0-alpha
INFO[0000] Retrieving image manifest 192.168.8.132:5000/wso2am/wso2micro-gw-toolkit:3.2.0-alpha
INFO[0000] Unpacking rootfs as cmd RUN micro-gw init project requires it.
INFO[0011] Taking snapshot of full filesystem...
INFO[0011] Resolving paths
INFO[0012] RUN micro-gw init project
INFO[0012] cmd: /bin/sh
INFO[0012] args: [-c micro-gw init project]
JAVA_HOME: /opt/java/openjdk
Initializing Toolkit...
Project 'project' is initialized successfully.

(Use "micro-gw build project" after copying the api definitions)
(Use "micro-gw import project [-l]|[-a -v]" to import APIs from WSO2 API Manager)
INFO[0014] Taking snapshot of full filesystem...
INFO[0014] Resolving paths
INFO[0017] RUN cp /usr/wso2/swagger/project-*/* ./project/api_definitions/
INFO[0017] cmd: /bin/sh
INFO[0017] args: [-c cp /usr/wso2/swagger/project-*/* ./project/api_definitions/]
INFO[0017] Taking snapshot of full filesystem...
INFO[0017] Resolving paths
INFO[0018] RUN cp /usr/wso2/policy/* ./project/
INFO[0018] cmd: /bin/sh
INFO[0018] args: [-c cp /usr/wso2/policy/* ./project/]
INFO[0018] Taking snapshot of full filesystem...
INFO[0018] Resolving paths
INFO[0019] RUN micro-gw build project
INFO[0019] cmd: /bin/sh
INFO[0019] args: [-c micro-gw build project]
JAVA_HOME: /opt/java/openjdk
Generating sources...[DONE]

..................................
There are no new updates available
..................................
Compiling source
	wso2/project:3.1.0
ballerina/java.jms:0.8.1 [central.ballerina.io -> home repo]  100% [==========
	ballerina/java.jms:0.8.1 pulled from central successfully

Creating balos
	target/balo/project-2020r1-java8-3.1.0.balo

Running Tests

	wso2/project:3.1.0
	No tests found


Generating executables
	target/bin/project.jar

BUILD SUCCESSFUL
Target: //project/target/project.jar
INFO[0057] Taking snapshot of full filesystem...
INFO[0057] Resolving paths
INFO[0058] RUN mkdir -p /home/exec
INFO[0058] cmd: /bin/sh
INFO[0058] args: [-c mkdir -p /home/exec]
INFO[0058] Taking snapshot of full filesystem...
INFO[0058] Resolving paths
INFO[0059] RUN cp ./project/target/*.jar /home/exec/
INFO[0059] cmd: /bin/sh
INFO[0059] args: [-c cp ./project/target/*.jar /home/exec/]
INFO[0059] Taking snapshot of full filesystem...
INFO[0059] Resolving paths
INFO[0060] Saving file /home/exec for later use
INFO[0060] Deleting filesystem...
INFO[0061] Retrieving image manifest 192.168.8.132:5000/wso2/wso2/wso2micro-gw:3.2.0-alpha
INFO[0061] Retrieving image manifest 192.168.8.132:5000/wso2/wso2/wso2micro-gw:3.2.0-alpha
INFO[0061] Unpacking rootfs as cmd RUN  mkdir -p /home/exec requires it.
INFO[0063] Taking snapshot of full filesystem...
INFO[0064] Resolving paths
INFO[0064] USER root
INFO[0064] cmd: USER
INFO[0064] RUN  mkdir -p /home/exec
INFO[0064] cmd: /bin/sh
INFO[0064] args: [-c mkdir -p /home/exec]
INFO[0064] Taking snapshot of full filesystem...
INFO[0064] Resolving paths
INFO[0064] COPY --from=toolkit /home/exec /home/exec
INFO[0064] Resolving paths
INFO[0064] Taking snapshot of files...
INFO[0064] RUN cp /usr/wso2/mgwconf/* /home/ballerina/conf
INFO[0064] cmd: /bin/sh
INFO[0064] args: [-c cp /usr/wso2/mgwconf/* /home/ballerina/conf]
INFO[0064] Taking snapshot of full filesystem...
INFO[0065] Resolving paths
INFO[0065] RUN echo yes | /home/ballerina/wso2/lib/jdk8*/bin/keytool -import  -storepass ballerina -keystore /home/ballerina/wso2/runtime/bre/security/ballerinaTruststore.p12 -alias "wso2am310-secret-security" -file /usr/wso2/certs/wso2am310-secret/server.pem
INFO[0065] cmd: /bin/sh
INFO[0065] args: [-c echo yes | /home/ballerina/wso2/lib/jdk8*/bin/keytool -import  -storepass ballerina -keystore /home/ballerina/wso2/runtime/bre/security/ballerinaTruststore.p12 -alias "wso2am310-secret-security" -file /usr/wso2/certs/wso2am310-secret/server.pem]
Certificate already exists in keystore under alias <apim300>
Do you still want to add it? [no]:  Certificate was added to keystore
INFO[0065] Taking snapshot of full filesystem...
INFO[0065] Resolving paths
INFO[0066] RUN chmod a+w /home/ballerina
INFO[0066] cmd: /bin/sh
INFO[0066] args: [-c chmod a+w /home/ballerina]
INFO[0066] Taking snapshot of full filesystem...
INFO[0066] Resolving paths
INFO[0066] USER ballerina
INFO[0066] cmd: USER
INFO[0066] ENV project="project"
```

**Comment by pubudu538** [comment]
This is fixed.
